src/libp2p/connection/established/multi_stream.rs

```diff
@@ -0,0 +1,757 @@
+                substream::Substream::ingoing(supported_protocols),
+                out_substream_id,
+            )
+        } else if self.ping_substream.is_none() {
+            let out_substream_id = self.next_out_substream_id;
+            self.next_out_substream_id += 1;
+
+            self.ping_substream = Some(id.clone());
+
+            (
+                substream::Substream::ping_out(self.ping_protocol.clone()),
+                out_substream_id,
+            )
+        } else if let Some(desired) = self.desired_out_substreams.pop_front() {
+            desired
+        } else {
+            // TODO: reset the new substream
+            todo!()
+        };
+
+        let previous_value = self.in_substreams.insert(id, (substream, out_substream_id));
+        if previous_value.is_some() {
+            // There is already a substream with that identifier. This is forbidden by the API of
+            // this function.
+            panic!()
+        }
+    }
+
+    /// Returns a list of substreams that the state machine would like to see reset. The user is
+    /// encouraged to call [`MultiStream::substream_read_write`] with this list of
+    /// substream.
+    ///
+    /// This value doesn't change automatically over time but only after a call to
+    /// [`MultiStream::substream_read_write`], [`MultiStream::inject_coordinator_message`],
+    /// [`MultiStream::add_substream`], or [`MultiStream::reset_substream`].
+    ///
+    /// This mechanism is not a substitute for [`ReadWrite::wake_up_after`]. If calling
+    /// [`MultiStream::substream_read_write`] puts something in [`ReadWrite::wake_up_after`], then
+    /// the substream should be processed after this time, even if it isn't return by this
+    /// method.
+    /// TODO: it should be the case ^ it is currently too complicated from the user's perspective to handle a `wake_up_after` value for each substream
+    ///
+    /// > **Note**: An example situation is: a notification is queued, which leads to a message
+    /// >           being sent to a connection task, which, once injected, leads to a notifications
+    /// >           substream being "ready" because it needs to send more data.
+    pub fn ready_substreams(&self) -> impl Iterator<Item = &TSubId> {
+        self.ready_substreams
+            .iter()
+            .filter_map(|in_id| self.out_in_substreams_map.get(in_id))
+    }
+
+    /// Immediately destroys the substream with the given identifier.
+    ///
+    /// The given identifier is now considered invalid by the state machine.
+    ///
+    /// # Panic
+    ///
+    /// Panics if there is no substream with that identifier.
+    ///
+    pub fn reset_substream(&mut self, substream_id: &TSubId) {
+        let (substream, out_substream_id) = self.in_substreams.remove(substream_id).unwrap();
+        self.ready_substreams.remove(&out_substream_id);
+
+        if Some(substream_id) == self.ping_substream.as_ref() {
+            self.ping_substream = None;
```

Comment:
If there's no ping substream open, then `desired_outbound_substreams` returns at least 1. The user is then supposed to call `add_substream`, into which if there's no ping substream open then it assigns the new substream for pinging.

After `reset_substream` is called, the user is supposed to call `desired_outbound_substreams` again. So if it's the ping substream that got reset, it will be reopened.